include/deal.II/lac/dynamic_sparsity_pattern.h

```diff
@@ -1010,15 +1070,21 @@ DynamicSparsityPattern::end (const size_type r) const
   Assert (r<n_rows(), ExcIndexRangeType<size_type>(r,0,n_rows()));
 
   // find the first row after r that has entries and return the begin
-  // iterator to it
+  // iterator to it. also skip rows for which we do not have
+  // store anything based on the IndexSet given to the sparsity
+  // pattern
   unsigned int row = r+1;
-  while ((row<n_rows()) && (row_length(row)==0))
-    ++row;
-
-  if (row == n_rows())
-    return iterator(this);
-  else
-    return iterator(this, row, 0);
+  while ((row<n_rows())
+         &&
+         ((row_length(row)==0)
+          ||
+          ((rowset.size() != 0) &&
+           (rowset.is_element(row) == false))))
+
+    if (row == n_rows())
+      return iterator(this);
+    else
+      return iterator(this, row, 0);
```

Comment:
Good question. I had only later realized that `row_length()` does the right thing for rows not locally stored. But you're completely right that the `++row` is missing somewhere. Patch forthcoming -- thanks for paying more attention that I apparently did!